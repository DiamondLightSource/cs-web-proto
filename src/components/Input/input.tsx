import React, { useState } from "react";
import { connectionWrapper } from "../ConnectionWrapper/connectionWrapper";
import { writePv } from "../../hooks/useCs";
import { VType } from "../../vtypes/vtypes";

import classes from "./input.module.css";
import { vtypeToString, stringToVtype } from "../../vtypes/utils";

export interface InputProps {
  pvName: string;
  value: string;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
  style?: object;
}

export const Input: React.FC<InputProps> = (props: InputProps): JSX.Element => (
  <input
    type="text"
    value={props.value}
    onKeyDown={props.onKeyDown}
    onChange={props.onChange}
    onBlur={props.onBlur}
    onClick={props.onClick}
    className={`Input ${classes.Input}`}
    style={props.style}
  />
);

interface ConnectedInputProps {
  pvName: string;
  style?: object;
}

interface SmartInputProps {
  pvName: string;
  value?: VType;
  style?: object;
}

export const SmartInput: React.FC<SmartInputProps> = (
  props: SmartInputProps
): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      writePv(props.pvName, stringToVtype(event.currentTarget.value));
      setInputValue("");
      setEditing(false);
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
    /* When focus lost show PV value. */
    setInputValue(vtypeToString(props.value));
  }

  if (!editing && inputValue !== vtypeToString(props.value)) {
    setInputValue(vtypeToString(props.value));
  }

  return (
    <Input
      pvName={props.pvName}
      value={inputValue}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
      style={props.style}
    />
  );
};

export const ConnectedInput: React.FC<ConnectedInputProps> = connectionWrapper(
  SmartInput
);
