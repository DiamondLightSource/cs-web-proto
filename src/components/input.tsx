import React, { useState } from "react";
import { useSelector } from "react-redux";
import { useSubscription, writePv } from "../hooks/useCs";
import { CsState } from "../redux/csState";

export interface InputProps {
  value: string;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
}

export const Input: React.FC<InputProps> = (props: InputProps): JSX.Element => (
  <input
    type="text"
    value={props.value}
    onKeyDown={props.onKeyDown}
    onChange={props.onChange}
    onBlur={props.onBlur}
    onClick={props.onClick}
  />
);

interface ConnectedInputProps {
  pvName: string;
}

export const ConnectedInput: React.FC<ConnectedInputProps> = (
  props: ConnectedInputProps
): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  useSubscription(props.pvName);
  /* It would be nice to be able to share a selector, but it's
     not immediately obvious how to make a selector that takes
     an argument other than state. */
  const latestValue: string = useSelector((state: CsState): string => {
    let value = state.valueCache[props.pvName];
    if (value == null) {
      return "";
    } else {
      return value.value.toString();
    }
  });
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      writePv(props.pvName, {
        type: "NTScalar",
        value: event.currentTarget.value
      });
      setInputValue("");
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
    /* When focus lost show PV value. */
    setInputValue(latestValue);
  }

  if (!editing && inputValue !== latestValue) {
    setInputValue(latestValue);
  }

  return (
    <Input
      value={inputValue}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
    />
  );
};
