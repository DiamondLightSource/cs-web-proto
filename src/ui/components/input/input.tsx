import React, { CSSProperties, useState } from "react";

export const InputComponent = (props: {
  value: string;
  onEnter: (value: string) => void;
  readonly?: boolean;
  style?: CSSProperties;
  className?: string;
}): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      props.onEnter(event.currentTarget.value);
      setInputValue("");
      setEditing(false);
      event.currentTarget.blur();
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!props.readonly && !editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
  }

  if (!editing && inputValue !== props.value) {
    setInputValue(props.value);
  }

  return (
    <input
      type="text"
      value={inputValue}
      readOnly={props.readonly}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
      className={props.className}
      style={props.style}
    />
  );
};
