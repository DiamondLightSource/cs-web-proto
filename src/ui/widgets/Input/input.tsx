import React, { useState } from "react";

import classes from "./input.module.css";
import { writePv } from "../../hooks/useSubscription";
import { commonCss, Widget } from "../widget";
import { PVInputComponent, PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  InferWidgetProps,
  FontPropOpt,
  ChoicePropOpt,
  ColorPropOpt,
  BoolPropOpt
} from "../propTypes";
import { Font } from "../../../types/font";
import { Color } from "../../../types/color";
import { AlarmQuality, DType } from "../../../types/dtypes";

export interface InputProps {
  pvName: string;
  value: string;
  readonly: boolean;
  foregroundColor?: Color;
  backgroundColor?: Color;
  transparent: boolean;
  alarm: AlarmQuality;
  alarmSensitive: boolean;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
  font?: Font;
  textAlign?: "left" | "center" | "right";
}

export const InputComponent: React.FC<InputProps> = (
  props: InputProps
): JSX.Element => {
  let allClasses = classes.Input;
  const style = commonCss(props);
  if (props.textAlign) {
    style.textAlign = props.textAlign;
  }
  style.color = props.foregroundColor?.toString();
  style.backgroundColor = props.backgroundColor?.toString();
  // Transparent prop overrides backgroundColor.
  if (props.transparent) {
    style["backgroundColor"] = "transparent";
  }
  if (props.readonly) {
    allClasses += ` ${classes.readonly}`;
  }
  if (props.alarmSensitive) {
    switch (props.alarm) {
      case AlarmQuality.UNDEFINED:
      case AlarmQuality.INVALID:
      case AlarmQuality.CHANGING:
        style.color = "var(--invalid)";
        break;
      case AlarmQuality.WARNING:
        style.color = "var(--alarm)";
        break;
      case AlarmQuality.ALARM:
        style.color = "var(--alarm)";
        break;
    }
  }
  return (
    <input
      type="text"
      value={props.value}
      onKeyDown={props.onKeyDown}
      onChange={props.onChange}
      onBlur={props.onBlur}
      onClick={props.onClick}
      className={allClasses}
      style={style}
      readOnly={props.readonly}
    />
  );
};

export const SmartInputComponent = (
  props: PVInputComponent & {
    font?: Font;
    foregroundColor?: Color;
    backgroundColor?: Color;
    transparent?: boolean;
    alarmSensitive?: boolean;
    textAlign?: "left" | "center" | "right";
  }
): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      writePv(
        props.pvName,
        new DType({ stringValue: event.currentTarget.value })
      );
      setInputValue("");
      setEditing(false);
      event.currentTarget.blur();
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!props.readonly && !editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
    /* When focus lost show PV value. */
    setInputValue(DType.coerceString(props.value));
  }

  if (!editing && inputValue !== DType.coerceString(props.value)) {
    setInputValue(DType.coerceString(props.value));
  }

  const alarmQuality = props.value?.getAlarm().quality ?? AlarmQuality.VALID;

  return (
    <InputComponent
      pvName={props.pvName}
      value={inputValue}
      alarm={alarmQuality}
      alarmSensitive={props.alarmSensitive || false}
      readonly={props.readonly}
      foregroundColor={props.foregroundColor}
      backgroundColor={props.backgroundColor}
      transparent={props.transparent ?? false}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
      font={props.font}
      textAlign={props.textAlign}
    />
  );
};

const InputWidgetProps = {
  ...PVWidgetPropType,
  font: FontPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  transparent: BoolPropOpt,
  alarmSensitive: BoolPropOpt,
  textAlign: ChoicePropOpt(["left", "center", "right"])
};

export const Input = (
  props: InferWidgetProps<typeof InputWidgetProps>
): JSX.Element => <Widget baseWidget={SmartInputComponent} {...props} />;

registerWidget(Input, InputWidgetProps, "input");
