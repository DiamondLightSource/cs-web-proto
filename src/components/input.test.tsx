import React from "react";
import { Input, InputProps } from "./input";
import { shallow } from "enzyme";

it("renders a basic element", (): void => {
  const mock = <jest.mock<typeof InputProps.onKeyDown>>InputProps.onKeyDown;
  const input = shallow(<Input 
    type="text"
    value="hello"
    onKeyDown={mock}
    onChange={mock}
    onBlur={mock}
    onClick={mock} />
    );
  expect(input.text()).toEqual("Hello");
});
