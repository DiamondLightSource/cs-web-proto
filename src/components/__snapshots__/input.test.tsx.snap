// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> it matches the snapshot 1`] = `
<input
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  type="text"
  value="hello"
/>
`;
